relay-replays/src/recording.rs

```diff
@@ -231,59 +235,72 @@ impl RecordingProcessor<'_> {
 /// -> META = 4
 /// -> CUSTOM = 5
 /// -> PLUGIN = 6
-
 #[derive(Debug, Serialize)]
 #[serde(untagged)]
-enum Event {
-    T2(FullSnapshotEvent),
-    T3(IncrementalSnapshotEvent),
-    T4(MetaEvent),
-    T5(CustomEvent),
+enum EventData {
+    FullSnapshot(FullSnapshotEventData),
+    IncrementalSnapshot(IncrementalSourceData),
+    Meta(Value),
+    Custom(CustomEventData),
     Default(Value),
-    // 0: DOMContentLoadedEvent,
-    // 1: LoadEvent,
-    // 6: PluginEvent,
+}
+
+#[derive(Debug, Serialize)]
+struct Event {
+    #[serde(rename = "type")]
+    ty: u8,
+    timestamp: f64,
+    data: EventData,
 }
 
 impl<'de> serde::Deserialize<'de> for Event {
     fn deserialize<D: serde::Deserializer<'de>>(d: D) -> Result<Self, D::Error> {
-        let value = Value::deserialize(d)?;
-
-        match value.get("type") {
-            Some(val) => match Value::as_u64(val) {
-                Some(v) => match v {
-                    2 => match FullSnapshotEvent::deserialize(value) {
-                        Ok(event) => Ok(Event::T2(event)),
-                        Err(_) => Err(DError::custom("could not parse snapshot event")),
-                    },
-                    3 => match IncrementalSnapshotEvent::deserialize(value) {
-                        Ok(event) => Ok(Event::T3(event)),
-                        Err(_) => Err(DError::custom("could not parse incremental snapshot event")),
-                    },
-                    4 => match MetaEvent::deserialize(value) {
-                        Ok(event) => Ok(Event::T4(event)),
-                        Err(_) => Err(DError::custom("could not parse meta event")),
-                    },
-                    5 => match CustomEvent::deserialize(value) {
-                        Ok(event) => Ok(Event::T5(event)),
-                        Err(e) => Err(DError::custom(e.to_string())),
-                    },
-                    0 | 1 | 6 => Ok(Event::Default(value)),
-                    _ => Err(DError::custom("invalid type value")),
-                },
-                None => Err(DError::custom("type field must be an integer")),
-            },
-            None => Err(DError::missing_field("type")),
+        #[derive(Deserialize)]
+        #[serde(bound(deserialize = "'de: 'a, 'a: 'de"))]
+        struct Helper<'a> {
+            #[serde(rename = "type")]
+            ty: u8,
+            timestamp: f64,
```

Comment:
@cmanallen what confused me in the existing impl: Some event types have timestamp `u64`, some have `f64`. Shouldn't they all have the same timestamp format?

Reply:
@jjbayer RRWeb timestamps are u64.  Sentry timestamps are f64.